crates/matrix-sdk-crypto/src/store/mod.rs

```diff
@@ -98,11 +98,84 @@ pub(crate) struct Store {
     inner: Arc<DynCryptoStore>,
     verification_machine: VerificationMachine,
     tracked_users_cache: Arc<DashSet<OwnedUserId>>,
-    users_for_key_query_cache: Arc<DashSet<OwnedUserId>>,
+    users_for_key_query: Arc<Mutex<UsersForKeyQuery>>,
     tracked_user_loading_lock: Arc<Mutex<()>>,
     tracked_users_loaded: Arc<AtomicBool>,
 }
 
+/// Record of the users that are waiting for a /keys/query.
+///
+/// To avoid races, we maintain a sequence number which is updated each time we
+/// receive an invalidation notification. We also record the sequence number at
+/// which each user was last invalidated. Then, we attach the current sequence
+/// number to each `/keys/query` request, and when we get the response we can
+/// tell if any users have been invalidated more recently than that request.
+#[derive(Debug)]
+struct UsersForKeyQuery {
+    /// the sequence number we will assign to the next addition to user_map
+    next_seqno: InvalidationSequenceNumber,
+
+    /// the users pending a lookup, together with the sequence number at which
+    /// they were added to the list
+    user_map: HashMap<OwnedUserId, InvalidationSequenceNumber>,
+}
+
+// we use a signed int for this, so that we can do a `wrapping_sub` and get a
+// meaningful result (which is ok even if it wraps)
+type InvalidationSequenceNumber = i64;
+
+impl UsersForKeyQuery {
+    /// create a new, empty, `UsersForKeyQueryCache`
+    fn new() -> Self {
+        UsersForKeyQuery { next_seqno: 0, user_map: HashMap::new() }
+    }
+
+    /// record a new user that requires a key query
+    fn insert_user(&mut self, user: &UserId) {
+        let seq = self.next_seqno;
+        trace!(?user, sequence_number = seq, "Flagging user for key query");
+        self.user_map.insert(user.to_owned(), seq);
+        self.next_seqno += 1;
+    }
+
+    /// record that a user has received an update with the given sequence
+    /// number.
+    ///
+    /// If the sequence number is newer than the oldest invalidation for this
+    /// user, it is removed from the list of those needing an update.
+    ///
+    /// Returns true if the user is now up-to-date, else false
+    fn maybe_remove_user(&mut self, user: &UserId, seqno: InvalidationSequenceNumber) -> bool {
+        let last_invalidation = self.user_map.get(user);
+        let _span = trace_span!(
+            "UsersForKeyQuery::maybe_remove_user",
+            ?user,
+            query_sequence = seqno,
+            invalidation_sequence = last_invalidation,
+        )
+        .entered();
+
+        if let Some(inv_seqno) = last_invalidation {
+            if inv_seqno.wrapping_sub(seqno) > 0 {
```

Comment:
If I understood this correctly this done so that you support an infinite amount of sequences instead of being limited to the range of an `u64`?

The comment on the type alias is a bit terse, perhaps we can expand a bit what an OK result is and why we are doing this. 

The mechanism also panics on debug builds since you also need to increment the `next_seqno` using wrapping addition.

Reply:
> If I understood this correctly this done so that you support an infinite amount of sequences instead of being limited to the range of an `u64`?
 
Yeah. Do you think it's overengineered? In a C world I'd do this because it's easier to make your sequence numbers wrap than to worry about whether 2**64 is enough, but since we have to be explicit in Rust, I'm not sure the same logic applies.

> The comment on the type alias is a bit terse, perhaps we can expand a bit what an OK result is and why we are doing this.
> 
> The mechanism also panics on debug builds since you also need to increment the `next_seqno` using wrapping addition.

Have fixed both of these.

Reply:
> Yeah. Do you think it's overengineered? In a C world I'd do this because it's easier to make your sequence numbers wrap than to worry about whether 2**64 is enough, but since we have to be explicit in Rust, I'm not sure the same logic applies.

I think we can leave it in, it's a nice property to have. Let's just document it a bit better. Oh and btw we could also have used an `u128`: https://doc.rust-lang.org/std/primitive.u128.html.

Reply:
We may need this kind of pattern in other places in the SDK by the way :-p.